core/src/main/java/com/netflix/iceberg/ReplaceFiles.java

```diff
@@ -16,21 +16,113 @@
 
 package com.netflix.iceberg;
 
-import java.util.List;
-import java.util.Set;
+import com.netflix.iceberg.exceptions.CommitFailedException;
+import com.netflix.iceberg.exceptions.RuntimeIOException;
+import com.netflix.iceberg.exceptions.ValidationException;
+import com.netflix.iceberg.io.OutputFile;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
 
 class ReplaceFiles extends SnapshotUpdate implements RewriteFiles {
-  public ReplaceFiles(TableOperations ops) {
-    super(ops);
-  }
 
-  @Override
-  protected List<String> apply(TableMetadata base) {
-    return null;
-  }
+    private final TableOperations ops;
+    private final Set<DataFile> filesToAdd = new HashSet<>();
+    private final Set<CharSequence> pathsToDelete = new HashSet<>();
+    private final List<String> newManifests = new ArrayList<>();
+    private List<String> cachedManifests = new ArrayList<>();
+    private final AtomicInteger manifestCount = new AtomicInteger(0);
+    private boolean hasChanges = false;
+
+    public ReplaceFiles(TableOperations ops) {
+        super(ops);
+        this.ops = ops;
+    }
+
+    /**
+     * @param base TableMetadata of the base snapshot
+     * @return list of manifests that "may" get committed if commit is called on this instance.
+     */
+    @Override
+    protected List<String> apply(TableMetadata base) {
+        ValidationException.check(!(this.pathsToDelete.isEmpty() || this.filesToAdd.isEmpty()), "Must provide some files to add and delete.");
+        if (this.hasChanges) {
+            final Snapshot snapshot = base.currentSnapshot();
+            ValidationException.check(snapshot != null, "No snapshots are committed so no datafiles can be replaced.");
+
+            final List<String> currentManifests = snapshot.manifests();
+            final List<CharSequence> deletedPaths = Collections.synchronizedList(new ArrayList<>());
+            final List<String> addedManifests = Collections.synchronizedList(new ArrayList<>());
+
+            currentManifests.parallelStream().forEach(manifest -> {
+                final OutputFile manifestPath = manifestPath(manifestCount.getAndIncrement());
+                try (ManifestReader manifestReader = ManifestReader.read(ops.newInputFile(manifest))) {
+                    try (ManifestWriter writer = new ManifestWriter(manifestReader.spec(), manifestPath, snapshotId())) {
+                        for (ManifestEntry manifestEntry : manifestReader.entries()) {
+                            if (this.pathsToDelete.contains(manifestEntry.file().path()))
+                                if (manifestEntry.status() == ManifestEntry.Status.DELETED) {
```

Comment:
If you scan use the reader as an `Interable` it will filter out the deleted entries for you. Then you just have to handle the case where some files weren't present in the manifest file.

Reply:
I need the manifest entry and there is no helper method for that right now. I am doing the filtering in this class now.